lib/trellis/plugins/vars/vars.py

```diff
@@ -20,14 +23,57 @@ def __init__(self, inventory):
         self.inventory_basedir = inventory.basedir()
         self._options = cli.options if cli else None
 
-    # Wrap salts and keys variables in {% raw %} to prevent jinja templating errors
-    def wrap_salts_in_raw(self, host, hostvars):
-        if 'vault_wordpress_sites' in hostvars:
-            for name, site in hostvars['vault_wordpress_sites'].iteritems():
-                for key, value in site['env'].iteritems():
-                    if key.endswith(('_key', '_salt')) and not value.startswith(('{% raw', '{%raw')):
-                        hostvars['vault_wordpress_sites'][name]['env'][key] = ''.join(['{% raw %}', value, '{% endraw %}'])
-            host.vars['vault_wordpress_sites'] = hostvars['vault_wordpress_sites']
+    def raw_triage(self, key_string, item, patterns):
+        # process dict values
+        if isinstance(item, AnsibleMapping):
+            dict = {}
+            for key,value in item.iteritems():
+                dict[key] = self.raw_triage('.'.join([key_string, key]), value, patterns)
+            return dict
+
+        # process list values
+        elif isinstance(item, AnsibleSequence):
+            list = []
+            for i,value in enumerate(item):
+                raw = self.raw_triage('.'.join([key_string, str(i)]), value, patterns)
+                list.append(raw)
+            return list
+
+        # wrap values if they match raw_vars pattern
+        elif isinstance(item, AnsibleUnicode):
+            matches = False
+            for pattern in patterns:
+                if re.match(pattern, key_string) is not None:
+                    matches = True
+                    break
+
+            if not item.startswith(('{% raw', '{%raw')) and matches:
+                item = ''.join(['{% raw %}', item, '{% endraw %}'])
+
+            return item
+
+    def raw_vars(self, host, hostvars):
+        if 'raw_vars' not in hostvars:
+            return
+
+        raw_vars = list((var for var in hostvars['raw_vars'] if var.split('.')[0] in hostvars))
+
+        # prepare regex match patterns
+        patterns = []
+        for pattern in raw_vars:
+            pattern = re.sub(r'\.', '\.', pattern)
+            pattern = re.sub(r'\*', '(.)*', pattern)
+            patterns.append(pattern)
+
+        # wrap matching vars under each key
+        keys = set()
+        for var in raw_vars:
+            key = var.split('.')[0]
+            if key in keys:
```

Comment:
The primary purpose of recursing only once through each top-level key is to avoid having multiple lists or dicts to merge for keys at various depths. No variable value is touched more than once. Each top-level key needs only one instance of the following and avoids merges:
```
host.vars[key] = self.raw_triage(key, hostvars[key], patterns)
```

A prior iteration had the potential to produce multiple dicts and lists per key, which I tried to merge using Ansible's [`merge_hash`](https://github.com/ansible/ansible/blob/v2.0.2.0-1/lib/ansible/utils/vars.py#L71_). There were edge-case merge failures, with nested values inappropriately overwitten (particularly for lists).

It is only a secondary benefit that we do no more work than for the simplest equivalent:
```yaml
raw_vars:
  - vault_foo
```